### PR TITLE
Log rangefinder signal quality

### PR DESCRIPTION
**Pull request by Williangalvani**

### CODE REVIEW

File: libraries/AP_Logger/LogStructure.h [hunk 1]
```diff
@@ -1076,6 +1077,7 @@ struct PACKED log_VER {
 // @Field: Stat: Sensor state
 // @FieldValueEnum: Stat: RangeFinder::Status
 // @Field: Orient: Sensor orientation
+// @Field: Quality: Signal quality. 0 is invalid, 1 is worst, 100 is best
```
Comment (peterbarker) [inline libraries/AP_Logger/LogStructure.h:1080]: Maybe make it `int8_t` and -1 is invalid?

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 2]
```diff
@@ -83,6 +83,16 @@ bool AP_RangeFinder_BLPing::get_reading(float &reading_m)
     return false;
 }
 
+bool AP_RangeFinder_BLPing::get_signal_quality_pct(uint8_t &quality_pct) const
+{
+    if (status() != RangeFinder::Status::Good) {
+        quality_pct = 0;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp:89]: Wasted flash - caller should be checking return value.

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 3]
```diff
@@ -83,6 +83,16 @@ bool AP_RangeFinder_BLPing::get_reading(float &reading_m)
...
+        return false;
+    }
+    quality_pct = MAX(1, protocol.get_confidence());
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp:92]: This doesn't look right - or a percentage.

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h [hunk 4]
```diff
@@ -133,6 +133,13 @@ class AP_RangeFinder_BLPing : public AP_RangeFinder_Backend_Serial
      */
     void update(void) override;
 
+    /**
+     * @brief Get the reading confidence as a MAVLink "Quality" signal
+     * 0 is invalid, 100 is best quality, 1 is worst quality
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h:138]: Let mavlink look after itself :-)  This method should be returning sensible thinks the mavlink library can interpret (which it does!)

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h [hunk 5]
```diff
@@ -133,6 +133,13 @@ class AP_RangeFinder_BLPing : public AP_RangeFinder_Backend_Serial
...
+     *
+     */
+    bool get_signal_quality_pct(uint8_t &quality_pct) const override;
```
Comment (Ryanf55) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h:141]: Add WARN_IF_UNUSED?

File: libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp [hunk 6]
```diff
@@ -27,13 +27,14 @@ void AP_RangeFinder_MAVLink::handle_msg(const mavlink_message_t &msg)
...
 
-    // only accept distances for the configured orentation
+    // only accept distances for the configured orientation
     if (packet.orientation == orientation()) {
         state.last_reading_ms = AP_HAL::millis();
         distance_cm = packet.current_distance;
         _max_distance_cm = packet.max_distance;
         _min_distance_cm = packet.min_distance;
         sensor_type = (MAV_DISTANCE_SENSOR)packet.type;
+        signal_quality = packet.signal_quality;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp:37]: This is not interpretting the mavlink field correctly:
```
      <field type="uint8_t" name="signal_quality" units="%">Signal quality of the sensor. Specific to each sensor type, representing the relation of the signal strength with the target reflectivity, distance, size or aspect, but normalised as a percentage. 0 = unknown/unset signal quality, 1 = invalid signal, 100 = perfect signal.</field>
```

File: libraries/AP_RangeFinder/AP_RangeFinder.cpp [hunk 7]
```diff
@@ -753,13 +753,18 @@ void RangeFinder::Log_RFND() const
             continue;
         }
 
+        uint8_t signal_quality = 0;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder.cpp:756]:
```suggestion
        int8_t signal_quality = 0;
```

etc

File: libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.h [hunk 8]
```diff
@@ -43,6 +47,7 @@ class AP_RangeFinder_MAVLink : public AP_RangeFinder_Backend
     uint16_t distance_cm;
     uint16_t _max_distance_cm;
     uint16_t _min_distance_cm;
+    uint8_t signal_quality;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.h:50]:
```suggestion
    int8_t signal_quality;
```

File: libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp [hunk 9]
```diff
@@ -77,4 +78,13 @@ void AP_RangeFinder_MAVLink::update(void)
     }
 }
 
+bool AP_RangeFinder_MAVLink::get_signal_quality_pct(uint8_t &quality_pct) const
+{
+    if (status() != RangeFinder::Status::Good) {
+        return false;
+    }
+    quality_pct = signal_quality;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp:86]:
```suggestion
    if (signal_quality <= 1) {
        return false;
    }
    quality_pct = signal_quality * 99 / 100;
```

not sure on that scaling, needs to scale from (1-100) to (0-100)

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp (hunk 6 above)
Comment (Williangalvani) [reply libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp:37]: so, signal quality 0 means mavlink has no data, so it maps to -1 in ardupilot, while 1 is the worst signal and is mapped to 0, does that make sense?

File: libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp (hunk 9 above)
Comment (Williangalvani) [reply libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp:93]: this is not done in handle_msg(), above

File: libraries/AP_RangeFinder/AP_RangeFinder.cpp [hunk 10]
```diff
@@ -768,13 +768,18 @@ void RangeFinder::Log_RFND() const
             continue;
         }
 
+        int8_t signal_quality = 0;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder.cpp:771]:
```suggestion
        int8_t signal_quality;
```

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h [hunk 11]
```diff
@@ -133,6 +133,13 @@ class AP_RangeFinder_BLPing : public AP_RangeFinder_Backend_Serial
      */
     void update(void) override;
 
+    /**
+     * @brief Get the reading confidence
+     * 100 is best quality, 0 is worst
+     *
+     */
+    WARN_IF_UNUSED bool get_signal_quality_pct(int8_t &quality_pct) const override;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h:141]:
```suggestion
    bool get_signal_quality_pct(int8_t &quality_pct) const override WARN_IF_UNUSED;
```

File: libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.h [hunk 12]
```diff
@@ -31,6 +31,10 @@ class AP_RangeFinder_MAVLink : public AP_RangeFinder_Backend
     int16_t max_distance_cm() const override;
     int16_t min_distance_cm() const override;
 
+    // Get the reading confidence
+    // 100 is best quality, 0 is worst
+    WARN_IF_UNUSED bool get_signal_quality_pct(int8_t &quality_pct) const override;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.h:36]:
```suggestion
    bool get_signal_quality_pct(int8_t &quality_pct) const override WARN_IF_UNUSED;
```

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 13]
```diff
@@ -394,7 +394,7 @@ void GCS_MAVLINK::send_distance_sensor(const AP_RangeFinder_Backend *sensor, con
         return;
     }
 
-    uint8_t quality_pct = 0;
+    int8_t quality_pct = 0;
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Common.cpp:397]:
```suggestion
    int8_t quality_pct;
```

**Review by peterbarker — CHANGES_REQUESTED** [review]
Looks pretty good

File: libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp [hunk 14]
```diff
@@ -27,13 +27,21 @@ void AP_RangeFinder_MAVLink::handle_msg(const mavlink_message_t &msg)
...
         _max_distance_cm = packet.max_distance;
         _min_distance_cm = packet.min_distance;
         sensor_type = (MAV_DISTANCE_SENSOR)packet.type;
+        signal_quality = packet.signal_quality;
+        if (signal_quality == 0) {
+            // MAVLink's 0 means invalid/unset, so we map it to -1
+            signal_quality = -1;
+        } else if (signal_quality == 1) {
+            // Map 1 to 0 as that is what ardupilot uses as the worst signal quality
+            signal_quality = 0;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp:43]: Should probably scale (x-1) *100/99 or something, but this will do...

**Pull request merged**